File: Leetcode/509. Fibonacci Number.py
# link - https://leetcode.com/problems/fibonacci-number/

class Solution:
    def fib(self, n: int) -> int:
        
        store = [None] * (n+1)
        
        def helper(n):
            if n < 2:
                return n
            if store[n]:
                return store[n]
            ans = helper(n-1) + helper(n-2)
            store[n] = ans
            return ans
        
        return helper(n)

File: Leetcode/509. Fibonacci Number.java

// Link - https://leetcode.com/problems/fibonacci-number/

// Spaace: O(1)
// Time: O(n)

class Solution {
    public int fib(int n) {
        
        if (n < 2) return n;
        
        int[] dp = {0, 1};
        int temp;
        
        for(int i=2; i < n; i++){
            
            temp = dp[1];
            dp[1] += dp[0];
            dp[0] = temp;
            
        }
        
        return dp[0] + dp[1];
        
    }
}